### PR TITLE
services/horizon/internal/db2/schema: Remove unnecessary tables and indexes

### PR DESCRIPTION
**Pull request by tamirms**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What
Close https://github.com/stellar/go/issues/2405

The asset_stats table is not used anymore because it has been replaced by exp_asset_stats.

index_history_transactions_on_id is not necessary because it is identical to hs_transaction_by_id

index_history_ledgers_on_id is not necessary because it is identical to hs_ledger_by_id

asset_by_code is not necessary because there is a compound index history_assets_asset_code_asset_type_asset_issuer_key
on `UNIQUE(asset_code, asset_type, asset_issuer)` which can be used on queries involving the `asset_code` column.

for the same reasons, exp_asset_stats_by_code can also be removed

### Why

These duplicate indexes don't improve query performance but do inhibit write performance. Removing the indexes will reduce storage and improve database write performance.

### Known limitations

N/A

### CODE REVIEW

**Comment by 2opremio** [comment]
> These duplicate indexes don't improve query performance but do inhibit write performance. Removing the indexes will reduce storage and improve database write performance.

Out of curiosity. Do you have an authoritative source on redundant indexes causing performance problems? I have been looking it up and couldn't find anything. Intuitively, I don't see why Postgres would optimize them out, but maybe that's the case.

File: services/horizon/internal/db2/schema/migrations/33_remove_unused.sql [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+DROP INDEX exp_asset_stats_by_code;
+
+DROP INDEX index_history_transactions_on_id;
+
+DROP INDEX index_history_ledgers_on_id;
+
+DROP INDEX asset_by_code;
+
+-- +migrate Down
```
Comment (2opremio) [inline services/horizon/internal/db2/schema/migrations/33_remove_unused.sql]: Where do we use backwards migration?

Comment (tamirms) [reply services/horizon/internal/db2/schema/migrations/33_remove_unused.sql]: there's a command called `horizon db migrate down <count>` which executes the backwards migration:

https://github.com/stellar/go/blob/master/services/horizon/cmd/db.go#L50

**Comment by tamirms** [comment]
@2opremio I couldn't find anything authoritative except for the "Duplicate indexes" section in https://wiki.postgresql.org/wiki/Index_Maintenance

However, I did run a SQL query on the horizon staging db to get index size and usage statistics. Here is the output: 

```
 schemaname |            tablename             |                       indexname                       |  num_rows   | table_size | index_size | unique | number_of_scans |  tuples_read  | tuples_fetched 
------------+----------------------------------+-------------------------------------------------------+-------------+------------+------------+--------+-----------------+---------------+----------------
 public     | exp_asset_stats                  | exp_asset_stats_pkey                                  |        9502 | 1416 kB    | 1560 kB    | Y      |          161001 |        449735 |         160990
 public     | exp_asset_stats                  | exp_asset_stats_by_code                               |        9502 | 1416 kB    | 408 kB     | N      |         3588799 |     569239833 |      568999115
 public     | history_assets                   | history_assets_asset_code_asset_type_asset_issuer_key |       26816 | 4920 kB    | 3840 kB    | Y      |        28632933 |      28632910 |       28632910
 public     | history_assets                   | asset_by_code                                         |       26816 | 4920 kB    | 792 kB     | N      |        15349015 |      25577560 |       25577560
 public     | history_ledgers                  | hs_ledger_by_id                                       | 2.67591e+07 | 20 GB      | 619 MB     | Y      |               0 |             0 |              0
 public     | history_ledgers                  | index_history_ledgers_on_id                           | 2.67591e+07 | 20 GB      | 619 MB     | Y      |       116066364 |     235943505 |      234242133
 public     | history_transactions             | hs_transaction_by_id                                  | 3.31465e+08 | 579 GB     | 7233 MB    | Y      |             102 |           102 |            102
 public     | history_transactions             | index_history_transactions_on_id                      | 3.31465e+08 | 579 GB     | 7233 MB    | Y      |    365093787345 |  150133108513 |   150015537260
```

If you look at the pairs (`hs_ledger_by_id`, `index_history_transactions_on_id`) and (`hs_transaction_by_id`, `index_history_transactions_on_id`) you'll find that the two indexes take up the same exact amount of space. Presumably, there must be time spent writing to the duplicate indexes. If we were to remove them then it would no longer be necessary to perform those write operations

File: services/horizon/internal/db2/schema/migrations/33_remove_unused.sql (hunk 1 above)
Comment (2opremio) [reply services/horizon/internal/db2/schema/migrations/33_remove_unused.sql]: Ah, that strikes me as odd. Is that actively used? (I couldn't find any references to it)

**Comment by 2opremio** [comment]
That makes sense. Thanks for the explanation . I wonder why `hs_transaction_by_id ` gets to be used instead of having a `0` counter.

Anyways, LGTM.

File: services/horizon/internal/db2/schema/migrations/33_remove_unused.sql (hunk 1 above)
Comment (tamirms) [reply services/horizon/internal/db2/schema/migrations/33_remove_unused.sql]: I don't think it's used very often but one use case I can think of is if you want to roll back to an earlier version of horizon. In that case, you may need to roll back some migrations as well

Comment (2opremio) [reply services/horizon/internal/db2/schema/migrations/33_remove_unused.sql]: Ah, really good point.

**Pull request merged**